File: 238-product-of-array-except-self/product-of-array-except-self.ts
function productExceptSelf(nums: number[]): number[] {
    let ans = new Array();
    let prefix = 1;
    ans[nums.length - 1] = 1;
    for (let i = nums.length - 1; i >= 0; i--) {
        if (i === nums.length - 1) ans[i] = nums[i];
        else ans[i] = ans[i + 1] * nums[i];
    }
    for (let i = 0; i < nums.length; i++) {
        if (i === nums.length - 1) ans[i] = prefix;
        else {
            ans[i] = prefix * ans[i + 1];
            prefix *= nums[i];
        }
    }
    return ans;
}

File: 238-product-of-array-except-self/product-of-array-except-self.java
class Solution {
    public int[] productExceptSelf(int[] nums) {
        int[] ans = new int[nums.length];
        int leftProduct = 1;
        for(int i = nums.length - 1; i >= 0; i --) {
            if(i == nums.length - 1) ans[i] = nums[i];
            else ans[i] = nums[i] * ans[i + 1];
        }
        for(int i = 0; i < nums.length; i ++) {
            if(i == nums.length - 1) ans[i] = leftProduct;
            else ans[i] = leftProduct * ans[i + 1];
            leftProduct *= nums[i];
        }
        return ans;
    }
}